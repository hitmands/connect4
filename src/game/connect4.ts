import * as R from 'ramda';
import { Game, Move } from './connect4.definitions';
import { evaluate } from './evaluate';

export const connect4 = (...moves: Move[]) => {
  const timeline = R.chain((move: Move) => [move, evaluate], moves);

  return (game: Game): Game => R.flow(game, timeline);
};
