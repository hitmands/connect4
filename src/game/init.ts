import { Cell, Game } from './connect4.definitions';
import { render } from './render';

export const toEmptyCell = (vector: Cell['vector']): Cell => ({
  player: null,
  vector,
});

export const toEmptyBoard = (): Game['board'] =>
  Array.from({ length: 6 }, (_, x) =>
    Array.from({ length: 7 }, (__, y) => toEmptyCell([x, y])),
  );

export const createGame = (): Game => ({
  history: [],
  board: toEmptyBoard(),
  toString() {
    return render(this);
  },
});
