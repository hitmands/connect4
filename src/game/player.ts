import * as R from 'ramda';
import { Cell, Game, Player, Vector } from './connect4.definitions';

export const invalid = R.assoc('error');

export const toPlayerColor = (p: unknown) => (p === '+' ? 'RED' : 'YELLOW');

export const turn = ({ history }: Game) =>
  history.length % 2 === 0 ? '+' : '-';

const findClosestAvailableVector = (column: number, { board }: Game) => {
  const x = board.findLastIndex((columns) => !columns[column].player);

  return x >= 0 ? ([x, column] as Vector) : undefined;
};

export const createPlayer =
  (player: Player) =>
  (column: number) =>
  (game: Game): Game => {
    if (game.winner || game.error) {
      return game;
    }

    if (turn(game) !== player) {
      const error = new TypeError(
        `It's not ${toPlayerColor(player)}'s (${player}) turn`,
      );

      return invalid(error, game);
    }

    const vector = findClosestAvailableVector(column, game);

    if (!vector) {
      const error = new RangeError(`Column "${column}" is full`);

      return invalid(error, game);
    }

    const cursor: Cell = {
      player,
      vector,
      date: new Date(),
    };

    return R.flow(game, [
      R.assocPath(['board', ...vector, 'player'], player),
      R.modify('history', R.prepend(cursor)),
    ]);
  };
