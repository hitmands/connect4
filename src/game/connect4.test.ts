import * as R from 'ramda';
import { describe, expect, it, test } from 'vitest';
import { connect4 } from './connect4';
import { Cell, Game } from './connect4.definitions';
import { createGame } from './init';
import { createPlayer, turn } from './player';

describe('Initializing the Game', () => {
  it('should draw a 6 rows by 7 columns empty board', () => {
    const game = createGame();

    expect(game.board).toHaveLength(6);
    expect(game.board).toSatisfy<Game['board']>((columns) =>
      columns.every((column) => column.length === 7),
    );

    expect(R.flow(game, [connect4(), R.toString])).toMatchInlineSnapshot(`
      "
      0|0|0|0|0|0|0
      0|0|0|0|0|0|0
      0|0|0|0|0|0|0
      0|0|0|0|0|0|0
      0|0|0|0|0|0|0
      0|0|0|0|0|0|0

      ------------------------------------------------------------
      STATUS:  | (+) RED MOVES NEXT;
      ------------------------------------------------------------
      HISTORY: | (0) Array<
               |  
               | >;
      "
    `);
  });

  it('should allow the red (+) the first move', () => {
    const game = createGame();

    expect(turn(game)).toEqual('+');
  });

  it('should initialize with an empty history', () => {
    const game = createGame();

    expect(game).toHaveProperty('history', []);
  });

  it('should initialize without a winner', () => {
    const game = createGame();

    expect(game).not.toHaveProperty('winner');
  });

  it('should initialize without an error', () => {
    const game = createGame();

    expect(game).not.toHaveProperty('error');
  });

  it('should contain a board of empty cells', () => {
    const game = createGame();

    expect(game).toHaveProperty('board');

    const assertion = ({ vector, player }: Cell) => {
      expect(player).toBeNull();

      // as indexes are 0-based
      const [x, y] = vector;

      expect(x).toBeGreaterThanOrEqual(0);
      expect(x).toBeLessThan(6);

      expect(y).toBeGreaterThanOrEqual(0);
      expect(y).toBeLessThan(7);
    };

    game.board.forEach((col) => col.forEach(assertion));
  });
});

describe('Playing the Game', () => {
  it('should only allow the red to move first', () => {
    const game = createGame();
    const red = createPlayer('+');
    const yellow = createPlayer('-');

    // The move is ignored
    const invalid = connect4(yellow(5));
    expect(invalid(game).toString()).toMatchInlineSnapshot(`
      "
      0|0|0|0|0|0|0
      0|0|0|0|0|0|0
      0|0|0|0|0|0|0
      0|0|0|0|0|0|0
      0|0|0|0|0|0|0
      0|0|0|0|0|0|0

      ------------------------------------------------------------
      STATUS:  | TypeError: It's not YELLOW's (-) turn;
      ------------------------------------------------------------
      HISTORY: | (0) Array<
               |  
               | >;
      "
    `);

    const valid = connect4(red(5));
    expect(valid(game).toString()).toMatchInlineSnapshot(`
      "
      0|0|0|0|0|0|0
      0|0|0|0|0|0|0
      0|0|0|0|0|0|0
      0|0|0|0|0|0|0
      0|0|0|0|0|0|0
      0|0|0|0|0|+|0

      ------------------------------------------------------------
      STATUS:  | (-) YELLOW MOVES NEXT;
      ------------------------------------------------------------
      HISTORY: | (1) Array<
               |  +
               | >;
      "
    `);
  });

  it('should allow pins to take the first available slot', () => {
    const game = createGame();
    const red = createPlayer('+');
    const yellow = createPlayer('-');

    const spec = connect4(
      red(5),
      yellow(5),
      red(4),
      yellow(5),
      red(4),
      yellow(3),
    );

    expect(spec(game).toString()).toMatchInlineSnapshot(`
      "
      0|0|0|0|0|0|0
      0|0|0|0|0|0|0
      0|0|0|0|0|0|0
      0|0|0|0|0|-|0
      0|0|0|0|+|-|0
      0|0|0|-|+|+|0

      ------------------------------------------------------------
      STATUS:  | (+) RED MOVES NEXT;
      ------------------------------------------------------------
      HISTORY: | (6) Array<
               |  +,-,+,-,+,-
               | >;
      "
    `);
  });

  it('should prevent the same player from playing twice in a row', () => {
    const game = createGame();
    const red = createPlayer('+');

    const spec = connect4(red(5), red(4));

    // The second move should be ignored
    expect(spec(game).toString()).toMatchInlineSnapshot(`
      "
      0|0|0|0|0|0|0
      0|0|0|0|0|0|0
      0|0|0|0|0|0|0
      0|0|0|0|0|0|0
      0|0|0|0|0|0|0
      0|0|0|0|0|+|0

      ------------------------------------------------------------
      STATUS:  | TypeError: It's not RED's (+) turn;
      ------------------------------------------------------------
      HISTORY: | (1) Array<
               |  +
               | >;
      "
    `);
  });

  it('should prevent a pin from being dropped on a full column', () => {
    const game = createGame();
    const red = createPlayer('+');
    const yellow = createPlayer('-');

    const spec = connect4(
      red(4),
      yellow(4),
      red(4),
      yellow(4),
      red(4),
      yellow(4),
      /* invalid */ red(4),
      /* ignored */ yellow(5),
    );

    expect(spec(game).toString()).toMatchInlineSnapshot(`
      "
      0|0|0|0|-|0|0
      0|0|0|0|+|0|0
      0|0|0|0|-|0|0
      0|0|0|0|+|0|0
      0|0|0|0|-|0|0
      0|0|0|0|+|0|0

      ------------------------------------------------------------
      STATUS:  | RangeError: Column "4" is full;
      ------------------------------------------------------------
      HISTORY: | (6) Array<
               |  +,-,+,-,+,-
               | >;
      "
    `);
  });
});

describe('Winning the Game', () => {
  it('should declare red the winner (5V)', () => {
    const game = createGame();
    const red = createPlayer('+');
    const yellow = createPlayer('-');

    const spec = connect4(
      red(5),
      yellow(3),
      red(5),
      yellow(3),
      red(5),
      yellow(2),
      red(5),
    );

    expect(spec(game).toString()).toMatchInlineSnapshot(`
      "
      0|0|0|0|0|0|0
      0|0|0|0|0|0|0
      0|0|0|0|0|+|0
      0|0|0|0|0|+|0
      0|0|0|-|0|+|0
      0|0|-|-|0|+|0

      ------------------------------------------------------------
      STATUS:  | (+) RED WINS,
               | Vertical(2,5 -> 3,5 -> 4,5 -> 5,5);
      ------------------------------------------------------------
      HISTORY: | (7) Array<
               |  +,-,+,-,+,-,+
               | >;
      "
    `);
  });

  it('should declare yellow the winner (0H)', () => {
    const game = createGame();
    const red = createPlayer('+');
    const yellow = createPlayer('-');

    const spec = connect4(
      red(0),
      yellow(2),
      red(1),
      yellow(3),
      red(3),
      yellow(4),
      red(3),
      yellow(5),
    );

    expect(spec(game).toString()).toMatchInlineSnapshot(`
      "
      0|0|0|0|0|0|0
      0|0|0|0|0|0|0
      0|0|0|0|0|0|0
      0|0|0|+|0|0|0
      0|0|0|+|0|0|0
      +|+|-|-|-|-|0

      ------------------------------------------------------------
      STATUS:  | (-) YELLOW WINS,
               | Horizontal(5,2 -> 5,3 -> 5,4 -> 5,5);
      ------------------------------------------------------------
      HISTORY: | (8) Array<
               |  +,-,+,-,+,-,+,-
               | >;
      "
    `);
  });

  it('should declare yellow the winner (1LD)', () => {
    const game = createGame();
    const red = createPlayer('+');
    const yellow = createPlayer('-');

    const spec = connect4(
      red(0),
      yellow(1),
      red(2),
      yellow(2),
      red(3),
      yellow(4),
      red(3),
      yellow(3),
      red(4),
      yellow(4),
      red(5),
      yellow(4),
    );

    expect(spec(game).toString()).toMatchInlineSnapshot(`
      "
      0|0|0|0|0|0|0
      0|0|0|0|0|0|0
      0|0|0|0|-|0|0
      0|0|0|-|-|0|0
      0|0|-|+|+|0|0
      +|-|+|+|-|+|0

      ------------------------------------------------------------
      STATUS:  | (-) YELLOW WINS,
               | CounterDiagonal(2,4 -> 3,3 -> 4,2 -> 5,1);
      ------------------------------------------------------------
      HISTORY: | (12) Array<
               |  +,-,+,-,+,-,+,-,+,-,
               |  +,-
               | >;
      "
    `);
  });

  it('should declare red the winner (0CD)', () => {
    const game = createGame();
    const red = createPlayer('+');
    const yellow = createPlayer('-');

    const spec = connect4(
      red(0),
      yellow(0),
      red(1),
      yellow(0),
      red(0),
      yellow(1),
      red(2),
      yellow(5),
      red(1),
      yellow(5),
      red(3),
    );

    expect(spec(game).toString()).toMatchInlineSnapshot(`
      "
      0|0|0|0|0|0|0
      0|0|0|0|0|0|0
      +|0|0|0|0|0|0
      -|+|0|0|0|0|0
      -|-|0|0|0|-|0
      +|+|+|+|0|-|0

      ------------------------------------------------------------
      STATUS:  | (+) RED WINS,
               | Horizontal(5,0 -> 5,1 -> 5,2 -> 5,3);
      ------------------------------------------------------------
      HISTORY: | (11) Array<
               |  +,-,+,-,+,-,+,-,+,-,
               |  +
               | >;
      "
    `);
  });
});

describe('Drawing the Game', () => {
  it('should declare it a draw', () => {
    const game = createGame();
    const red = createPlayer('+');
    const yellow = createPlayer('-');

    const spec = connect4(
      red(0),
      yellow(0),
      red(0),
      yellow(0),
      red(0),
      yellow(0),

      red(1),
      yellow(1),
      red(1),
      yellow(1),
      red(1),
      yellow(1),

      red(3),
      yellow(2),
      red(2),
      yellow(2),
      red(2),
      yellow(2),

      red(4),
      yellow(2),
      red(3),
      yellow(3),
      red(3),
      yellow(3),

      red(3),
      yellow(4),
      red(4),
      yellow(4),
      red(5),
      yellow(5),

      red(4),
      yellow(4),
      red(5),
      yellow(5),
      red(5),
      yellow(6),
      red(5),
      yellow(6),
      red(6),
      yellow(6),
      red(6),
      yellow(6),
      red(6),
    );

    expect(spec(game).toString()).toMatchInlineSnapshot(`
      "
      -|-|-|+|-|+|-
      +|+|-|-|+|+|+
      -|-|+|+|-|-|-
      +|+|-|-|+|+|+
      -|-|+|+|-|-|-
      +|+|-|+|+|+|-

      ------------------------------------------------------------
      STATUS:  | RangeError: DRAW;
      ------------------------------------------------------------
      HISTORY: | (42) Array<
               |  +,-,+,-,+,-,+,-,+,-,
               |  +,-,+,-,+,-,+,-,+,-,
               |  +,-,+,-,+,-,+,-,+,-,
               |  +,-,+,-,+,-,+,-,+,-,
               |  +,-
               | >;
      "
    `);
  });
});

describe('fixtures', async () => {
  const { default: fixtures } = await import('./fixtures.json');

  const table = test.each(fixtures);

  table('%#', ({ moves }) => {
    const game = createGame();
    const red = createPlayer('+');
    const yellow = createPlayer('-');

    const spec = connect4(
      ...moves.map((col, i) => (i % 2 === 0 ? red : yellow)(col)),
    );

    expect(spec(game).toString()).toMatchSnapshot();
  });
});
