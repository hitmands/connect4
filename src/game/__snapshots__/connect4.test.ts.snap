// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fixtures > 0 1`] = `
"
0|0|0|0|0|0|0
0|0|0|0|0|0|0
+|0|0|0|0|0|0
+|-|0|0|0|0|0
+|-|0|0|0|0|0
+|-|0|0|0|0|0

------------------------------------------------------------
STATUS:  | (+) RED WINS,
         | Vertical(2,0 -> 3,0 -> 4,0 -> 5,0);
------------------------------------------------------------
HISTORY: | (7) Array<
         |  +,-,+,-,+,-,+
         | >;
"
`;

exports[`fixtures > 1 1`] = `
"
0|0|0|0|0|0|0
0|0|0|0|0|0|0
0|+|0|0|0|0|0
0|+|-|0|0|0|0
0|+|-|0|0|0|0
0|+|-|0|0|0|0

------------------------------------------------------------
STATUS:  | (+) RED WINS,
         | Vertical(2,1 -> 3,1 -> 4,1 -> 5,1);
------------------------------------------------------------
HISTORY: | (7) Array<
         |  +,-,+,-,+,-,+
         | >;
"
`;

exports[`fixtures > 2 1`] = `
"
0|0|0|0|0|0|0
0|0|0|0|0|0|0
0|0|+|0|0|0|0
0|0|+|-|0|0|0
0|0|+|-|0|0|0
0|0|+|-|0|0|0

------------------------------------------------------------
STATUS:  | (+) RED WINS,
         | Vertical(2,2 -> 3,2 -> 4,2 -> 5,2);
------------------------------------------------------------
HISTORY: | (7) Array<
         |  +,-,+,-,+,-,+
         | >;
"
`;

exports[`fixtures > 3 1`] = `
"
0|0|0|0|0|0|0
0|0|0|0|0|0|0
0|0|0|+|0|0|0
0|0|0|+|-|0|0
0|0|0|+|-|0|0
0|0|0|+|-|0|0

------------------------------------------------------------
STATUS:  | (+) RED WINS,
         | Vertical(2,3 -> 3,3 -> 4,3 -> 5,3);
------------------------------------------------------------
HISTORY: | (7) Array<
         |  +,-,+,-,+,-,+
         | >;
"
`;

exports[`fixtures > 4 1`] = `
"
0|0|0|0|0|0|0
0|0|0|0|0|0|0
0|0|0|0|+|0|0
0|0|0|0|+|-|0
0|0|0|0|+|-|0
0|0|0|0|+|-|0

------------------------------------------------------------
STATUS:  | (+) RED WINS,
         | Vertical(2,4 -> 3,4 -> 4,4 -> 5,4);
------------------------------------------------------------
HISTORY: | (7) Array<
         |  +,-,+,-,+,-,+
         | >;
"
`;

exports[`fixtures > 5 1`] = `
"
0|0|0|0|0|0|0
0|0|0|0|0|0|0
0|0|0|0|0|+|0
0|0|0|0|0|+|-
0|0|0|0|0|+|-
0|0|0|0|0|+|-

------------------------------------------------------------
STATUS:  | (+) RED WINS,
         | Vertical(2,5 -> 3,5 -> 4,5 -> 5,5);
------------------------------------------------------------
HISTORY: | (7) Array<
         |  +,-,+,-,+,-,+
         | >;
"
`;

exports[`fixtures > 6 1`] = `
"
0|0|0|0|0|0|0
0|0|0|0|0|0|0
0|0|0|0|0|0|+
-|0|0|0|0|0|+
-|0|0|0|0|0|+
-|0|0|0|0|0|+

------------------------------------------------------------
STATUS:  | (+) RED WINS,
         | Vertical(2,6 -> 3,6 -> 4,6 -> 5,6);
------------------------------------------------------------
HISTORY: | (7) Array<
         |  +,-,+,-,+,-,+
         | >;
"
`;

exports[`fixtures > 7 1`] = `
"
0|0|0|0|0|0|0
0|0|0|0|0|0|0
-|0|0|0|0|0|0
+|-|+|-|+|-|+
-|+|-|+|-|+|-
+|-|+|-|+|-|+

------------------------------------------------------------
STATUS:  | (-) YELLOW WINS,
         | LeadDiagonal(2,0 -> 3,1 -> 4,2 -> 5,3);
------------------------------------------------------------
HISTORY: | (22) Array<
         |  +,-,+,-,+,-,+,-,+,-,
         |  +,-,+,-,+,-,+,-,+,-,
         |  +,-
         | >;
"
`;

exports[`fixtures > 8 1`] = `
"
0|0|0|0|0|0|0
0|0|0|0|0|0|0
0|0|0|0|0|0|-
+|-|+|-|+|-|+
-|+|-|+|-|+|-
+|-|+|-|+|-|+

------------------------------------------------------------
STATUS:  | (-) YELLOW WINS,
         | CounterDiagonal(2,6 -> 3,5 -> 4,4 -> 5,3);
------------------------------------------------------------
HISTORY: | (22) Array<
         |  +,-,+,-,+,-,+,-,+,-,
         |  +,-,+,-,+,-,+,-,+,-,
         |  +,-
         | >;
"
`;

exports[`fixtures > 9 1`] = `
"
0|0|0|0|0|0|0
0|0|0|0|0|0|0
0|0|0|0|0|0|0
0|0|0|0|0|0|0
-|-|-|0|0|0|0
+|+|+|+|0|0|0

------------------------------------------------------------
STATUS:  | (+) RED WINS,
         | Horizontal(5,0 -> 5,1 -> 5,2 -> 5,3);
------------------------------------------------------------
HISTORY: | (7) Array<
         |  +,-,+,-,+,-,+
         | >;
"
`;

exports[`fixtures > 10 1`] = `
"
0|0|0|0|0|0|0
0|0|0|0|0|0|0
0|0|0|0|0|0|0
0|0|0|0|0|0|0
0|0|0|0|-|-|-
0|0|0|+|+|+|+

------------------------------------------------------------
STATUS:  | (+) RED WINS,
         | Horizontal(5,3 -> 5,4 -> 5,5 -> 5,6);
------------------------------------------------------------
HISTORY: | (7) Array<
         |  +,-,+,-,+,-,+
         | >;
"
`;

exports[`fixtures > 11 1`] = `
"
0|0|0|0|0|0|0
0|0|0|0|0|0|0
0|0|0|0|0|0|-
+|-|+|-|+|-|+
-|+|-|+|-|+|-
+|-|+|-|+|-|+

------------------------------------------------------------
STATUS:  | (-) YELLOW WINS,
         | CounterDiagonal(2,6 -> 3,5 -> 4,4 -> 5,3);
------------------------------------------------------------
HISTORY: | (22) Array<
         |  +,-,+,-,+,-,+,-,+,-,
         |  +,-,+,-,+,-,+,-,+,-,
         |  +,-
         | >;
"
`;

exports[`fixtures > 12 1`] = `
"
0|0|0|0|0|0|0
0|0|0|0|0|0|0
-|0|0|0|0|0|0
+|-|+|-|+|-|+
-|+|-|+|-|+|-
+|-|+|-|+|-|+

------------------------------------------------------------
STATUS:  | (-) YELLOW WINS,
         | LeadDiagonal(2,0 -> 3,1 -> 4,2 -> 5,3);
------------------------------------------------------------
HISTORY: | (22) Array<
         |  +,-,+,-,+,-,+,-,+,-,
         |  +,-,+,-,+,-,+,-,+,-,
         |  +,-
         | >;
"
`;

exports[`fixtures > 13 1`] = `
"
0|0|0|0|0|0|0
0|0|0|0|0|0|0
0|0|0|0|0|0|0
0|0|0|0|0|0|0
-|-|-|0|0|0|0
+|+|+|+|0|0|0

------------------------------------------------------------
STATUS:  | (+) RED WINS,
         | Horizontal(5,0 -> 5,1 -> 5,2 -> 5,3);
------------------------------------------------------------
HISTORY: | (7) Array<
         |  +,-,+,-,+,-,+
         | >;
"
`;

exports[`fixtures > 14 1`] = `
"
0|0|0|0|0|0|0
0|0|0|0|0|0|0
0|0|0|0|0|0|0
0|0|0|0|0|0|0
0|0|0|0|-|-|-
0|0|0|+|+|+|+

------------------------------------------------------------
STATUS:  | (+) RED WINS,
         | Horizontal(5,3 -> 5,4 -> 5,5 -> 5,6);
------------------------------------------------------------
HISTORY: | (7) Array<
         |  +,-,+,-,+,-,+
         | >;
"
`;

exports[`fixtures > 15 1`] = `
"
0|0|0|0|0|0|0
0|0|0|0|0|0|0
0|0|0|0|0|0|-
+|-|+|-|+|-|+
-|+|-|+|-|+|-
+|-|+|-|+|-|+

------------------------------------------------------------
STATUS:  | (-) YELLOW WINS,
         | CounterDiagonal(2,6 -> 3,5 -> 4,4 -> 5,3);
------------------------------------------------------------
HISTORY: | (22) Array<
         |  +,-,+,-,+,-,+,-,+,-,
         |  +,-,+,-,+,-,+,-,+,-,
         |  +,-
         | >;
"
`;

exports[`fixtures > 16 1`] = `
"
0|0|0|0|0|0|0
0|0|0|0|0|0|0
-|0|0|0|0|0|0
+|-|+|-|+|-|+
-|+|-|+|-|+|-
+|-|+|-|+|-|+

------------------------------------------------------------
STATUS:  | (-) YELLOW WINS,
         | LeadDiagonal(2,0 -> 3,1 -> 4,2 -> 5,3);
------------------------------------------------------------
HISTORY: | (22) Array<
         |  +,-,+,-,+,-,+,-,+,-,
         |  +,-,+,-,+,-,+,-,+,-,
         |  +,-
         | >;
"
`;

exports[`fixtures > 17 1`] = `
"
0|0|0|0|0|0|0
0|0|0|0|0|0|0
0|0|0|0|0|0|0
0|0|0|0|0|0|0
-|-|-|0|0|0|0
+|+|+|+|0|0|0

------------------------------------------------------------
STATUS:  | (+) RED WINS,
         | Horizontal(5,0 -> 5,1 -> 5,2 -> 5,3);
------------------------------------------------------------
HISTORY: | (7) Array<
         |  +,-,+,-,+,-,+
         | >;
"
`;

exports[`fixtures > 18 1`] = `
"
0|0|0|0|0|0|0
0|0|0|0|0|0|0
0|0|0|0|0|0|0
0|0|0|0|0|0|0
0|0|0|0|-|-|-
0|0|0|+|+|+|+

------------------------------------------------------------
STATUS:  | (+) RED WINS,
         | Horizontal(5,3 -> 5,4 -> 5,5 -> 5,6);
------------------------------------------------------------
HISTORY: | (7) Array<
         |  +,-,+,-,+,-,+
         | >;
"
`;

exports[`fixtures > 19 1`] = `
"
0|0|0|0|0|0|0
0|0|0|0|0|0|0
0|0|0|0|0|0|-
+|-|+|-|+|-|+
-|+|-|+|-|+|-
+|-|+|-|+|-|+

------------------------------------------------------------
STATUS:  | (-) YELLOW WINS,
         | CounterDiagonal(2,6 -> 3,5 -> 4,4 -> 5,3);
------------------------------------------------------------
HISTORY: | (22) Array<
         |  +,-,+,-,+,-,+,-,+,-,
         |  +,-,+,-,+,-,+,-,+,-,
         |  +,-
         | >;
"
`;

exports[`fixtures > 20 1`] = `
"
0|0|0|0|0|0|0
0|0|0|0|0|0|0
-|0|0|0|0|0|0
+|-|+|-|+|-|+
-|+|-|+|-|+|-
+|-|+|-|+|-|+

------------------------------------------------------------
STATUS:  | (-) YELLOW WINS,
         | LeadDiagonal(2,0 -> 3,1 -> 4,2 -> 5,3);
------------------------------------------------------------
HISTORY: | (22) Array<
         |  +,-,+,-,+,-,+,-,+,-,
         |  +,-,+,-,+,-,+,-,+,-,
         |  +,-
         | >;
"
`;
